File: src/main/java/dev/verloren/midnight/highlighter/CompactSyntaxHighlighter.java
package dev.verloren.midnight.highlighter;

import com.intellij.lexer.Lexer;


import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IElementType;
import dev.verloren.midnight.lexer.CompactLexer;
import dev.verloren.midnight.lexer.CompactTokenTypes;
import org.jspecify.annotations.NonNull;

/**
 * Syntax highlighter for Compact source code.
 *
 * <p>Maps lexical tokens from {@link CompactLexer} to IntelliJ {@link TextAttributesKey}s
 * for editor syntax highlighting.</p>
 */
public class CompactSyntaxHighlighter extends SyntaxHighlighterBase {

  public static final TextAttributesKey KEYWORD = CompactHighlighterColors.KEYWORD;
  public static final TextAttributesKey TYPE = CompactHighlighterColors.BUILTIN_TYPE;
  public static final TextAttributesKey STRING = CompactHighlighterColors.STRING;
  public static final TextAttributesKey NUMBER = CompactHighlighterColors.NUMBER;
  public static final TextAttributesKey COMMENT = CompactHighlighterColors.LINE_COMMENT;
  public static final TextAttributesKey BAD_CHARACTER = CompactHighlighterColors.BAD_CHARACTER;
  public static final TextAttributesKey RESERVED_KEYWORD = CompactHighlighterColors.RESERVED_KEYWORD;
  public static final TextAttributesKey CONSTANT = CompactHighlighterColors.BOOLEAN;
  public static final TextAttributesKey VERSION = CompactHighlighterColors.VERSION;
  public static final TextAttributesKey OPERATOR = CompactHighlighterColors.OPERATOR;
  public static final TextAttributesKey PARENTHESES = CompactHighlighterColors.PARENTHESES;
  public static final TextAttributesKey BRACES = CompactHighlighterColors.BRACES;
  public static final TextAttributesKey BRACKETS = CompactHighlighterColors.BRACKETS;
  public static final TextAttributesKey COMMA = CompactHighlighterColors.COMMA;
  public static final TextAttributesKey SEMICOLON = CompactHighlighterColors.SEMICOLON;
  public static final TextAttributesKey DOT = CompactHighlighterColors.DOT;
  public static final TextAttributesKey COLON = CompactHighlighterColors.COLON;
  public static final TextAttributesKey PRAGMA = CompactHighlighterColors.PRAGMA;
  public static final TextAttributesKey DOC_COMMENT = CompactHighlighterColors.DOC_COMMENT;
  public static final TextAttributesKey BLOCK_COMMENT = CompactHighlighterColors.BLOCK_COMMENT;

  @Override
  public @NonNull Lexer getHighlightingLexer() {
    return new CompactLexer();
  }

  @Override
  public TextAttributesKey @NonNull [] getTokenHighlights(IElementType tokenType) {
    if (tokenType == CompactTokenTypes.EXPORT ||
        tokenType == CompactTokenTypes.PURE ||
        tokenType == CompactTokenTypes.SEALED ||
        tokenType == CompactTokenTypes.NEW ||
        tokenType == CompactTokenTypes.IMPLEMENTS ||
        tokenType == CompactTokenTypes.EXTERNAL) {
      return pack(CompactHighlighterColors.MODIFIER);
    }

    if (tokenType == CompactTokenTypes.PRAGMA) {
      return pack(PRAGMA);
    }

    if (tokenType == CompactTokenTypes.ASSERT ||
        tokenType == CompactTokenTypes.DISCLOSE ||
        tokenType == CompactTokenTypes.FOLD ||
        tokenType == CompactTokenTypes.SLICE ||
        tokenType == CompactTokenTypes.PAD ||
        tokenType == CompactTokenTypes.EMIT ||
        tokenType == CompactTokenTypes.MAP) {
      return pack(CompactHighlighterColors.BUILTIN_FUNCTION);
    }

    if (
            tokenType == CompactTokenTypes.IMPORT ||
            tokenType == CompactTokenTypes.FROM ||
            tokenType == CompactTokenTypes.MODULE ||
            tokenType == CompactTokenTypes.PREFIX ||
            tokenType == CompactTokenTypes.AS ||
            tokenType == CompactTokenTypes.CIRCUIT ||
            tokenType == CompactTokenTypes.CONST ||
            tokenType == CompactTokenTypes.CONSTRUCTOR ||
            tokenType == CompactTokenTypes.CONTRACT ||
            tokenType == CompactTokenTypes.DEFAULT ||
            tokenType == CompactTokenTypes.ELSE ||
            tokenType == CompactTokenTypes.ENUM ||
            tokenType == CompactTokenTypes.FOR ||
            tokenType == CompactTokenTypes.IF ||
            tokenType == CompactTokenTypes.INCLUDE ||
            tokenType == CompactTokenTypes.LEDGER ||
            tokenType == CompactTokenTypes.OF ||
            tokenType == CompactTokenTypes.RETURN ||
            tokenType == CompactTokenTypes.STRUCT ||
            tokenType == CompactTokenTypes.TYPE ||
            tokenType == CompactTokenTypes.WITNESS ||
            tokenType == CompactTokenTypes.HASH ||
            tokenType == CompactTokenTypes.LET
    ) {
      return pack(KEYWORD);
    }

    if (tokenType == CompactTokenTypes.BOOLEAN_TYPE ||
            tokenType == CompactTokenTypes.BYTES_TYPE ||
            tokenType == CompactTokenTypes.FIELD_TYPE ||
            tokenType == CompactTokenTypes.OPAQUE_TYPE ||
            tokenType == CompactTokenTypes.UINT_TYPE ||
            tokenType == CompactTokenTypes.VECTOR_TYPE ||
            tokenType == CompactTokenTypes.JUBJUB_SCALAR_TYPE ||
            tokenType == CompactTokenTypes.SECP256K1_BASE_TYPE ||
            tokenType == CompactTokenTypes.SECP256K1_SCALAR_TYPE) {
      return pack(TYPE);
    }

    if (tokenType == CompactTokenTypes.TRUE || tokenType == CompactTokenTypes.FALSE) {
      return pack(CONSTANT);
    }

    if (tokenType == CompactTokenTypes.STRING_LITERAL) {
      return pack(STRING);
    }

    if (tokenType == CompactTokenTypes.VERSION_LITERAL) {
      return pack(VERSION);
    }

    if (tokenType == CompactTokenTypes.DECIMAL_LITERAL ||
            tokenType == CompactTokenTypes.BINARY_LITERAL ||
            tokenType == CompactTokenTypes.OCTAL_LITERAL ||
            tokenType == CompactTokenTypes.HEX_LITERAL) {
      return pack(NUMBER);
    }

    if (tokenType == CompactTokenTypes.BLOCK_COMMENT) {
      return pack(BLOCK_COMMENT);
    }
    if (tokenType == CompactTokenTypes.LINE_COMMENT) {
      return pack(COMMENT);
    }

    if (
            tokenType == CompactTokenTypes.PLUS ||
                    tokenType == CompactTokenTypes.MINUS ||
                    tokenType == CompactTokenTypes.STAR ||
                    tokenType == CompactTokenTypes.SLASH ||
                    tokenType == CompactTokenTypes.PERCENT ||
                    tokenType == CompactTokenTypes.ASSIGN ||
                    tokenType == CompactTokenTypes.EQEQ ||
                    tokenType == CompactTokenTypes.NEQ ||
                    tokenType == CompactTokenTypes.LTE ||
                    tokenType == CompactTokenTypes.GTE ||
                    tokenType == CompactTokenTypes.LT ||
                    tokenType == CompactTokenTypes.GT ||
                    tokenType == CompactTokenTypes.NOT ||
                    tokenType == CompactTokenTypes.ANDAND ||
                    tokenType == CompactTokenTypes.OROR ||
                    tokenType == CompactTokenTypes.ARROW ||
                    tokenType == CompactTokenTypes.SPREAD ||
                    tokenType == CompactTokenTypes.RANGE ||
                    tokenType == CompactTokenTypes.PLUS_ASSIGN ||
                    tokenType == CompactTokenTypes.MINUS_ASSIGN ||
                    tokenType == CompactTokenTypes.QUESTION
    ) {
      return pack(OPERATOR);
    }

    if (tokenType == CompactTokenTypes.LPAREN || tokenType == CompactTokenTypes.RPAREN) {
      return pack(PARENTHESES);
    }

    if (tokenType == CompactTokenTypes.LBRACE || tokenType == CompactTokenTypes.RBRACE) {
      return pack(BRACES);
    }
    if (tokenType == CompactTokenTypes.LBRACKET || tokenType == CompactTokenTypes.RBRACKET) {
      return pack(BRACKETS);
    }

    if (tokenType == CompactTokenTypes.COMMA) {
      return pack(COMMA);
    }

    if (tokenType == CompactTokenTypes.COLON) {
      return pack(COLON);
    }

    if (tokenType == CompactTokenTypes.SEMICOLON) {
      return pack(SEMICOLON);
    }

    if (tokenType == CompactTokenTypes.DOT) {
      return pack(DOT);
    }

    if (tokenType == CompactTokenTypes.UNTERMINATED_STRING ||
            tokenType == CompactTokenTypes.UNTERMINATED_BLOCK_COMMENT ||
            tokenType == TokenType.BAD_CHARACTER) {
      return pack(BAD_CHARACTER);
    }

    return new TextAttributesKey[0];
  }
}

File: src/main/kotlin/dev/verloren/midnight/highlighter/CompactSyntaxHighlighter.kt
package dev.verloren.midnight.highlighter

import com.intellij.lexer.Lexer
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
import com.intellij.openapi.editor.HighlighterColors
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IElementType
import dev.verloren.midnight.lexer.CompactLexerAdapter
import dev.verloren.midnight.lexer.CompactTokenTypes

class CompactSyntaxHighlighter : SyntaxHighlighterBase() {

    override fun getHighlightingLexer(): Lexer = CompactLexerAdapter()

    override fun getTokenHighlights(tokenType: IElementType): Array<TextAttributesKey> {
        return pack(
            when (tokenType) {

                /* ===================================================== */
                /* Keywords */
                /* ===================================================== */

                CompactTokenTypes.EXPORT,
                CompactTokenTypes.IMPORT,
                CompactTokenTypes.FROM,
                CompactTokenTypes.MODULE,
                CompactTokenTypes.PREFIX,
                CompactTokenTypes.ASSERT,
                CompactTokenTypes.AS,
                CompactTokenTypes.CIRCUIT,
                CompactTokenTypes.CONST,
                CompactTokenTypes.CONSTRUCTOR,
                CompactTokenTypes.CONTRACT,
                CompactTokenTypes.DEFAULT,
                CompactTokenTypes.DISCLOSE,
                CompactTokenTypes.ELSE,
                CompactTokenTypes.ENUM,
                CompactTokenTypes.FOLD,
                CompactTokenTypes.FOR,
                CompactTokenTypes.IF,
                CompactTokenTypes.INCLUDE,
                CompactTokenTypes.LEDGER,
                CompactTokenTypes.MAP,
                CompactTokenTypes.NEW,
                CompactTokenTypes.OF,
                CompactTokenTypes.PAD,
                CompactTokenTypes.PRAGMA,
                CompactTokenTypes.PURE,
                CompactTokenTypes.RETURN,
                CompactTokenTypes.SEALED,
                CompactTokenTypes.SLICE,
                CompactTokenTypes.STRUCT,
                CompactTokenTypes.TYPE,
                CompactTokenTypes.WITNESS,
                CompactTokenTypes.EMIT ->
                    KEYWORD

                CompactTokenTypes.RESERVED_KEYWORD ->
                    RESERVED_KEYWORD

                /* ===================================================== */
                /* Types */
                /* ===================================================== */

                CompactTokenTypes.BOOLEAN_TYPE,
                CompactTokenTypes.BYTES_TYPE,
                CompactTokenTypes.FIELD_TYPE,
                CompactTokenTypes.OPAQUE_TYPE,
                CompactTokenTypes.UINT_TYPE,
                CompactTokenTypes.VECTOR_TYPE,
                CompactTokenTypes.JUBJUB_SCALAR_TYPE,
                CompactTokenTypes.SECP256K1_BASE_TYPE,
                CompactTokenTypes.SECP256K1_SCALAR_TYPE ->
                    TYPE

                /* ===================================================== */
                /* Constants */
                /* ===================================================== */

                CompactTokenTypes.TRUE,
                CompactTokenTypes.FALSE ->
                    CONSTANT

                /* ===================================================== */
                /* Literals */
                /* ===================================================== */

                CompactTokenTypes.STRING_LITERAL ->
                    STRING

                CompactTokenTypes.VERSION_LITERAL ->
                    VERSION

                CompactTokenTypes.DECIMAL_LITERAL,
                CompactTokenTypes.BINARY_LITERAL,
                CompactTokenTypes.OCTAL_LITERAL,
                CompactTokenTypes.HEX_LITERAL ->
                    NUMBER

                /* ===================================================== */
                /* Comments */
                /* ===================================================== */

                CompactTokenTypes.LINE_COMMENT,
                CompactTokenTypes.BLOCK_COMMENT ->
                    COMMENT

                /* ===================================================== */
                /* Operators */
                /* ===================================================== */

                CompactTokenTypes.PLUS,
                CompactTokenTypes.MINUS,
                CompactTokenTypes.STAR,
                CompactTokenTypes.SLASH,
                CompactTokenTypes.PERCENT,
                CompactTokenTypes.ASSIGN,
                CompactTokenTypes.EQEQ,
                CompactTokenTypes.NEQ,
                CompactTokenTypes.LTE,
                CompactTokenTypes.GTE,
                CompactTokenTypes.LT,
                CompactTokenTypes.GT,
                CompactTokenTypes.NOT,
                CompactTokenTypes.ANDAND,
                CompactTokenTypes.OROR,
                CompactTokenTypes.ARROW,
                CompactTokenTypes.SPREAD,
                CompactTokenTypes.RANGE,
                CompactTokenTypes.PLUS_ASSIGN,
                CompactTokenTypes.MINUS_ASSIGN ->
                    OPERATOR

                /* ===================================================== */
                /* Delimiters */
                /* ===================================================== */

                CompactTokenTypes.LPAREN,
                CompactTokenTypes.RPAREN ->
                    PARENTHESES

                CompactTokenTypes.LBRACE,
                CompactTokenTypes.RBRACE ->
                    BRACES

                CompactTokenTypes.LBRACKET,
                CompactTokenTypes.RBRACKET ->
                    BRACKETS

                CompactTokenTypes.COMMA ->
                    COMMA

                CompactTokenTypes.SEMICOLON ->
                    SEMICOLON

                CompactTokenTypes.DOT ->
                    DOT

                /* ===================================================== */
                /* Errors */
                /* ===================================================== */

                CompactTokenTypes.UNTERMINATED_STRING,
                CompactTokenTypes.UNTERMINATED_BLOCK_COMMENT,
                TokenType.BAD_CHARACTER ->
                    BAD_CHARACTER

                else -> null
            }
        )
    }

    companion object {
        val KEYWORD =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_KEYWORD",
                DefaultLanguageHighlighterColors.KEYWORD
            )

        val TYPE =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_TYPE",
                DefaultLanguageHighlighterColors.CLASS_NAME
            )

        val STRING =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_STRING",
                DefaultLanguageHighlighterColors.STRING
            )

        val NUMBER =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_NUMBER",
                DefaultLanguageHighlighterColors.NUMBER
            )

        val COMMENT =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_COMMENT",
                DefaultLanguageHighlighterColors.LINE_COMMENT
            )

        val BAD_CHARACTER =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_BAD_CHARACTER",
                HighlighterColors.BAD_CHARACTER
            )

        val RESERVED_KEYWORD =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_RESERVED_KEYWORD",
                DefaultLanguageHighlighterColors.KEYWORD
            )

        val CONSTANT =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_CONSTANT",
                DefaultLanguageHighlighterColors.CONSTANT
            )

        val VERSION =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_VERSION",
                DefaultLanguageHighlighterColors.NUMBER
            )

        val OPERATOR =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_OPERATOR",
                DefaultLanguageHighlighterColors.OPERATION_SIGN
            )

        val PARENTHESES =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_PARENTHESES",
                DefaultLanguageHighlighterColors.PARENTHESES
            )

        val BRACES =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_BRACES",
                DefaultLanguageHighlighterColors.BRACES
            )

        val BRACKETS =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_BRACKETS",
                DefaultLanguageHighlighterColors.BRACKETS
            )

        val COMMA =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_COMMA",
                DefaultLanguageHighlighterColors.COMMA
            )

        val SEMICOLON =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_SEMICOLON",
                DefaultLanguageHighlighterColors.SEMICOLON
            )

        val DOT =
            TextAttributesKey.createTextAttributesKey(
                "COMPACT_DOT",
                DefaultLanguageHighlighterColors.DOT
            )
    }
}